torch/csrc/autograd/autograd.h

```diff
@@ -75,5 +75,25 @@ TORCH_API variable_list grad(
     bool create_graph = false,
     bool allow_unused = false);
 
+namespace forward_ad {
+
+/// Creates a new dual level and returns its index. This level index should then be used to call
+/// into the other functions below.
+TORCH_API uint64_t enter_dual_level();
+
+/// Exits the given level. This will clear up all the gradients from this level and all dual Tensors
+/// that had gradients for this level will become regular Tensors again.
+TORCH_API void exit_dual_level(uint64_t level);
+
+/// This function can be used to create a dual Tensor that holds a tangent to compute forward mode gradients.
+/// Note that the dual Tensor's primal is a view of the given primal and the given tangent is used as-is.
+/// This function is backward differentiable.
+TORCH_API at::Tensor make_dual(const at::Tensor& primal, at::Tensor tangent, uint64_t level);
+/// This function can be used to unpack a given dual Tensor to get its primal and tangent. The returned primal
+/// is a view of the dual and the tangent is returned as is.
+/// This function is backward differentiable.
+TORCH_API std::pair<at::Tensor, at::Tensor> unpack_dual(const at::Tensor& tensor, uint64_t level);
```

Comment:
How come these aren't functions in `native_functions.yaml`? Would save you from having to bind them manually.

Reply:
Can the second of this pair be undefined? If so, would it be better to do `c10::optional<at::Tensor>` here?

Reply:
This has been moved to be proper operators.